### PR TITLE
fix(zone.js): fesm2015 bundle should also be strict module.

### LINKED ISSUE #40215: The ngOnDestroy handler crashes after upgrade to Angular 11 on 'ng test' (zone.js issue)
# 🐞 bug report

### Affected Package
The issue is caused by package zone.js

### Is this a regression?

Yes, the previous version in which this bug was not present was: Angular 10

### Description

After updating Angular to version 11 and using the latest zone.js (0.11.3),  on `ng test` the ngOnDestroy handler crashes throwing:
```
ERROR Error: Uncaught (in promise): TypeError: Cannot read property 'data-ck-expando' of undefined
TypeError: Cannot read property 'data-ck-expando' of undefined
```

It is because `Object.prototype.toString.call( undefined )` returns `[Object Window]` instead of `[Object Undefined]` when the strict mode isn't turned on.

A similar issue was already reporter in the past: https://github.com/angular/angular/issues/31678

The workaround is to import zone.js in `polyfills.js` like this `import 'zone.js/dist/zone.js'` instead of `import 'zone.js/dist/zone'` or `import 'zone.js'`

## 🌍  Your Environment

<pre>
@angular-devkit/architect       0.1100.4
@angular-devkit/build-angular   0.1100.4
@angular-devkit/core            11.0.4
@angular-devkit/schematics      11.0.4
@schematics/angular             11.0.4
@schematics/update              0.1100.4
ng-packagr                      11.0.3
rxjs                            6.6.3
typescript                      4.0.3
webpack                         4.44.2
</pre>

### PR DESCRIPTION
**Pull request by JiaLiPassion**
Close #40215

`fesm2015/zone.js` is built to `esm` bundle with rollup, so the 'use strict';
statement is not generated in the bundle, even we put the 'use strict' in the src code,
rollup removes the code in the final bundle.

So if we load the `fesm2015/zone.js` as a module, such as `ng serve`, in the index.html

```
<script src="polyfills.js" type="module"></script>
```

Everything works fine, since polyfills.js is loaded as `module`, so it is always `strict`.

But in `ng test`, webpack concat the `zone.js` and loaded into the karma html. For other app and
 test code, they are still `strict` since they are `module` because they have `export/import`
statement, but `zone.js` is a bundle without `export`, it is a `side effect` bundle, so after
 loaded by webpack, it becomes non-strict. Which causes some issues, such as #40215,
the root cause is the `this` context should be `undefined` but treated as `Window` in `non-strict` mode.

```
Object.prototype.toString.apply(undefined);
// should be [object undefined], but it is [object Window] in non-strict mode.
```

So this commit always add `'use strict';` statement at the beginning of the zone.js es2015 bundle.

### CODE REVIEW

**Review by alan-agius4 — COMMENTED** [review]
LGTM.

Although I think it's simpler to just add it as part of the banner.  https://github.com/angular/angular/blob/4e8198d60f421ce120e3a6b57afe60a9332d2692/packages/zone.js/rollup-es5_global-es2015.config.js#L18

The fact that rollup is removing the use strict pragma makes sense since ES modules are strict by default. 

However the problem here is that when used with Webpack this is causing an issue. Webpack will typically add the use-strict pragma to ES modules, however Webpack is unable to determine that zone.js is fact an ES module because it doesn't contain any expressions such as import or export statements.

**Review by alan-agius4 — APPROVED** [review]
LGTM.

Although I think it's simpler to just add it as part of the banner.  https://github.com/angular/angular/blob/4e8198d60f421ce120e3a6b57afe60a9332d2692/packages/zone.js/rollup-es5_global-es2015.config.js#L18

The fact that rollup is removing the use strict pragma makes sense since ES modules are strict by default. 

However the problem here is that when used with Webpack this is causing an issue. Webpack will typically add the use-strict pragma to ES modules, however Webpack is unable to determine that zone.js is fact an ES module because it doesn't contain any expressions such as import or export statements.

File: packages/zone.js/test/npm_package/npm_package.spec.ts [hunk 1]
```diff
@@ -86,6 +86,12 @@ describe('Zone.js npm_package', () => {
           expect(shx.cat('zone.umd.js')).not.toContain('sourceMappingURL');
         });
       });
+
+      it('zone.js(es5) should contain use strict', () => {
+        checkInSubFolder('./bundles', () => {
+          expect(shx.cat('zone.umd.js')).toContain('use strict');
```
Comment (gkalpak) [inline packages/zone.js/test/npm_package/npm_package.spec.ts:92]: Consider making this test (and the es2015 one) more strict by checking for `'use strict';` and possibly that the file starts with that. Just `toContain('use strict')` could easily catch other occurrences of the words, which are not related to the `'use strict'` pragma.

For example:

```suggestion
          expect(shx.cat('zone.umd.js')).toMatch(/^\s*'use strict';/);
```

**Review by gkalpak — APPROVED** [review]
LGTM for `dev-infra`.
I can't really speak about the implementation, but @alan-agius4's suggestion of putting it in the banner sounds reasonable (and less "magic") to me.

Reviewed-for: dev-infra

**Comment by JiaLiPassion** [comment]
@alan-agius4, @gkalpak , thanks for the review. I have updated the PR.
Adding `use strict` to the `license banner` is much simpler, but it has another issue that is the `es5` bundle will have double `use strict` generated.

such as `bundles/zone.js` will look like

```
'use strict';
/**
* @license Angular v<unknown>
* (c) 2010-2020 Google LLC. https://angular.io/
* License: MIT
*/
(function (factory) {
    typeof define === 'function' && define.amd ? define(factory) :
        factory();
}((function () {
    'use strict';
```

Which should not be a real issue.

File: packages/zone.js/test/npm_package/npm_package.spec.ts (hunk 1 above)
Comment (JiaLiPassion) [reply packages/zone.js/test/npm_package/npm_package.spec.ts]: Thanks, I have updated this one.

**Review by alan-agius4 — APPROVED** [review]

File: packages/zone.js/rollup-es5.config.js [hunk 2]
```diff
@@ -14,7 +14,16 @@ if (bazel_stamp_file) {
...
+// because zone.js is a module without export and import, webpack is unable
+// to determine the bundle is `esm` module or not, so it doesn't add the 'use strict'
+// which webpack does to all other `esm` modules which has export or import.
+// And it causes issues such as https://github.com/angular/angular/issues/40215
+// `this` should be `undefined` but is assigned with `Window` instead.
+const banner = `'use strict';
+/**
 * @license Angular v${version}
 * (c) 2010-2020 Google LLC. https://angular.io/
 * License: MIT
```
Comment (gkalpak) [inline packages/zone.js/rollup-es5.config.js:29]: Not directly related to this PR, but these lines are missing a leading space :grin:

File: packages/zone.js/rollup-es5_global-es2015.config.js [hunk 3]
```diff
@@ -14,7 +14,16 @@ if (bazel_stamp_file) {
...
+// because zone.js is a module without export and import, webpack is unable
+// to determine the bundle is `esm` module or not, so it doesn't add the 'use strict'
+// which webpack does to all other `esm` modules which has export or import.
+// And it causes issues such as https://github.com/angular/angular/issues/40215
+// `this` should be `undefined` but is assigned with `Window` instead.
+const banner = `'use strict';
+/**
 * @license Angular v${version}
 * (c) 2010-2020 Google LLC. https://angular.io/
 * License: MIT
```
Comment (gkalpak) [inline packages/zone.js/rollup-es5_global-es2015.config.js:29]: Not directly related to this PR, but these lines are missing a leading space 😁

File: packages/zone.js/rollup-es5.config.js [hunk 4]
```diff
@@ -14,7 +14,16 @@ if (bazel_stamp_file) {
...
-const banner = `/**
+// Add 'use strict' to the bundle, https://github.com/angular/angular/pull/40456
+// When rollup build esm bundle of zone.js, there will be no 'use strict'
+// since all esm bundles are `strict`, but when webpack load the esm bundle,
+// because zone.js is a module without export and import, webpack is unable
+// to determine the bundle is `esm` module or not, so it doesn't add the 'use strict'
+// which webpack does to all other `esm` modules which has export or import.
+// And it causes issues such as https://github.com/angular/angular/issues/40215
+// `this` should be `undefined` but is assigned with `Window` instead.
+const banner = `'use strict';
```
Comment (gkalpak) [inline packages/zone.js/rollup-es5.config.js:25]: Do we really need this for the es5 config? AFAICT, we are no longer publishing ESM formats for ES5.

Comment (alan-agius4) [reply packages/zone.js/rollup-es5.config.js:25]: Good point :)

File: packages/zone.js/rollup-es5.config.js (hunk 2 above)
Comment (JiaLiPassion) [reply packages/zone.js/rollup-es5.config.js:29]: Got it, thanks, I will update the comment here.

File: packages/zone.js/rollup-es5_global-es2015.config.js (hunk 3 above)
Comment (JiaLiPassion) [reply packages/zone.js/rollup-es5_global-es2015.config.js:29]: Got it, thanks, I will update the comment here.

File: packages/zone.js/rollup-es5.config.js (hunk 4 above)
Comment (JiaLiPassion) [reply packages/zone.js/rollup-es5.config.js:25]: The name of this file is a little confusing, I will make a refactor later, this `rollup` config is not only for `es5` bundle, but also for `esm` bundle, the difference between `rollup-es5` and `rollup-es2015` is the definitions of `global`, and the to output `esm` or `es5` is controlled by `bazel` downstream target, so in fact, I need to rename the `rollup config` and make a common one, I will do it in another PR. Thanks.

**Pull request closed without merging**